Sofa/Component/Mapping/NonLinear/src/sofa/component/mapping/nonlinear/VolumeMapping.inl

```diff
@@ -0,0 +1,368 @@
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+#include <sofa/component/mapping/nonlinear/VolumeMapping.h>
+#include <sofa/core/BaseLocalMappingMatrix.h>
+#include <sofa/linearalgebra/CompressedRowSparseMatrixConstraintEigenUtils.h>
+
+namespace sofa::component::mapping::nonlinear
+{
+
+using sofa::type::dyad;
+using sofa::type::dot;
+
+template <class TIn, class TOut>
+VolumeMapping<TIn, TOut>::VolumeMapping()
+    : l_topology(initLink("topology", "link to the topology container"))
+{}
+
+template <class TIn, class TOut>
+auto VolumeMapping<TIn, TOut>::computeSecondDerivativeVolume(
+    const sofa::type::fixed_array<sofa::type::Vec3, 4>& tetrahedronVertices) ->
+    sofa::type::Mat<4, 4, sofa::type::Mat<3, 3, Real> >
+{
+    const sofa::type::fixed_array<sofa::type::Vec3, 3> v {
+        tetrahedronVertices[1] - tetrahedronVertices[0],
+        tetrahedronVertices[2] - tetrahedronVertices[0],
+        tetrahedronVertices[3] - tetrahedronVertices[0]
+    };
+
+    const auto H12 = sofa::type::crossProductMatrix(v[1] - v[2]) / 6;
+    const auto H13 = sofa::type::crossProductMatrix(v[2] - v[0]) / 6;
+    const auto H14 = sofa::type::crossProductMatrix(v[0] - v[1]) / 6;
+
+    const auto H23 = sofa::type::crossProductMatrix(-v[2]) / 6;
+    const auto H24 = sofa::type::crossProductMatrix(v[1]) / 6;
+    const auto H34 = sofa::type::crossProductMatrix(-v[0]) / 6;
+
+    sofa::type::Mat<4, 4, sofa::type::Mat<3, 3, SReal> > hessian;
+
+    hessian(0, 1) = H12;
+    hessian(1, 0) = H12.transposed();
+    hessian(0, 2) = H13;
+    hessian(2, 0) = H13.transposed();
+    hessian(0, 3) = H14;
+    hessian(3, 0) = H14.transposed();
+    hessian(1, 2) = H23;
+    hessian(2, 1) = H23.transposed();
+    hessian(1, 3) = H24;
+    hessian(3, 1) = H24.transposed();
+    hessian(2, 3) = H34;
+    hessian(3, 2) = H34.transposed();
+
+    return hessian;
+}
+
+template <class TIn, class TOut>
+void VolumeMapping<TIn, TOut>::init()
+{
+    if (l_topology.empty())
+    {
+        msg_warning() << "link to Topology container should be set to ensure right behavior. First Topology found in current context will be used.";
+        l_topology.set(this->getContext()->getMeshTopologyLink());
+    }
+
+    msg_info() << "Topology path used: '" << l_topology.getLinkedPath() << "'";
+
+    const auto nbTetrahedra = l_topology->getNbTetrahedra();
+
+    if (l_topology->getNbTetrahedra() == 0)
+    {
+        msg_error() << "No topology component containing tetrahedron found at path: " << l_topology.getLinkedPath() << ", nor in current context: " << this->getContext()->name;
+        this->d_componentState.setValue(sofa::core::objectmodel::ComponentState::Invalid);
+        return;
+    }
```

Comment:
Volume could be hexas as well, so I guess it should be tested and warned that it is not supported

Reply:
I trigger an error and set the component state to invalid. Just like in `AreaMapping`